src/linear_algebra.jl

```diff
@@ -12,15 +12,63 @@ corresponding to the assignment is as large as possible under lexicographical or
 Return a matrix of Bools where each column has a single true in it corresponding to its
 assignment.
 """
-function closest_permutation(mat::AbstractMatrix{<:Real})
-    ans = zeros(Bool, size(mat)...) # initialize answer as all false
-    queue = collect(1:size(mat, 2)) # columns to be assigned
-    while length(queue) > 0
-        col = pop!(queue)
-        possible_assignments = collect(1:size(mat, 1)) # row indices
+function closest_permutation(mat::AbstractMatrix)
+    nrows, ncols = size(mat)
+    @assert nrows >= ncols
+    ans = similar(BitArray, axes(mat)) # initialize answer as all false
+    fill!(ans, false)
+
+    ax2 = axes(mat, 2)
+
+    # Do a multi-threaded first pass, ignoring collisions.
+    colidxs = similar(ax2)
+    Threads.@threads for col in ax2
+        @inbounds begin
+            _, i = findmax(mat[:, col])
+            ans[i, col] = true
+            colidxs[col] = i
+        end
+    end
+
+    # Prepare a BitArray `collisions` and mark collided columns in it.
+    sp = sortperm(colidxs)
+    sortedidxs = colidxs[sp]
+    collisions = similar(BitArray, ax2)
+    fill!(collisions, false)
+
+    previndex = firstindex(sortedidxs)
+    @inbounds for i in eachindex(sortedidxs)
+        isequal(i, firstindex(sortedidxs)) && continue
+        if isequal(sortedidxs[i], sortedidxs[previndex])
+            collisions[sp[i]] = true
+            collisions[sp[previndex]] = true
+        end
+        previndex = i
+    end
+
+    @inbounds stack = sp[collisions]
```

Comment:
nice decorator, i didn't know about it

Reply:
I'm pretty wary of using `@inbounds` unless I'm absolutely sure I'm using the indexing interface generically, given that it can crash Julia if used improperly, but you can see I have taken pains to avoid `for i in 1:length(array)` and other indexing anti-patterns.